[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #84423 [8/N] [Dispatchable Collectives] Update allgather with CPU / CUDA implementations
* #83916 [7/N] [Dispatchable Collectives] Update reduce with CPU / CUDA implementations
* #83876 [6/N] [Dispatchable Collectives] Update recv with CPU / CUDA implementations
* #83859 [5/N] [Dispatchable Collectives] Update send with CPU / CUDA implementations
* **#83810 [4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations**
* #83735 [3/N] [Dispatchable Collectives] Update broadcast_ with CPU and CUDA implementations

### About this PR
* Update the all_reduce op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
* Update test to validate that a separate device implementation is not supported.

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D39506979](https://our.internmc.facebook.com/intern/diff/D39506979)

H-Huang committed:
[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations

[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/OpsImpl.cpp:
```diff
@@ -42,6 +68,24 @@ TORCH_LIBRARY_IMPL(c10d, CPU, m) {
+// sparse all_reduce in the Gloo backend
+TORCH_LIBRARY_IMPL(c10d, SparseCPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
```
nit: is there actually a `SparseCPU` or `SparseCUDA` implementation in Gloo?
If not, do we need to add it here?

H-Huang replied:
There is logic within `allreduce()` to handle sparse tensors (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L1465-L1467) to which it branches to different logic, so having each implementation just call allreduce_ keeps behavior the same, but I think there may be a cleaner way to do this, we should discuss this.

kwen2501 reviewed (approved):
LGTM!

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"




[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39506979).

kumpera commented on torch/csrc/distributed/c10d/OpsImpl.cpp:
```diff
@@ -49,6 +84,24 @@ TORCH_LIBRARY_IMPL(c10d, CPU, m) {
+  m.impl("allreduce_", allreduce_cpu_);
+}
+
+TORCH_LIBRARY_IMPL(c10d, SparseCUDA, m) {
```
Why do we have allreduce_cuda_ and allreduce_cpu_ when they look identical?

H-Huang replied:
Right now is setting up the skeleton to have dispatching route to different implementations for CPU/CUDA. Yes the implementations are identical for now, they will be updated in the future to perform `process_group->backend[<cpu>/<cuda>]->all_reduce`, but right now it is a no-op.

H-Huang committed:
Update on "[4/N] [Dispatchable Collectives] Update all_reduce_ with CPU / CUDA implementations"


### About this PR
* Update the all_reduce op to dispatch to cpu and cuda implementations. Right now they both perform the same logic so this is essentially a no-op.
* Update test to validate that a separate device implementation is not supported.

### About this stack
In the future we will repurpose ProcessGroup to instead contain a list of Backends (ProcessGroupNCCL/Gloo/UCC) and perform dispatching to them based on tensor type. The CPU and CUDA implementations will be updated to have process group select its CPU and CUDA backends respectively.

Differential Revision: [D39506979](https://our.internmc.facebook.com/intern/diff/D39506979)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39506979).

H-Huang commented:
@pytorchbot merge
